[11/n][FlexiRaft] Implementation of static modes in FlexiRaft

iRitwik opened the pull request:
Contains the following commits:

**[FlexiRaft][kuduraft] Quorum simplification**
This change modifies the configuration to denote the modes called out in the note.

**[7/n][FlexiRaft] Leader election quorum optimization**
This change implements the quorum optimizer when the election is not conducted in a term immediately succeeding that of the last known leader.

**[10/n][FlexiRaft] Simplification of data commit flow for dynamic quorum mode**
Simplifies code for the data commit flow as per the new modes.

**[11/n][FlexiRaft] Implementation of static modes in FlexiRaft**
Implementation of the static modes.

anirbanr-fb commented:
extensive Offline review. We discussed the key improvements as the next  steps.

1. We will make voter_distribution not dependent on Active  Config but passed in expected voters list.
2. Make the PotentialNextLeader response be able to jump across terms to a term where we had decidability
3. Majority of Majority regions for SINGLE_REGION_DYNAMIC

iRitwik commented:
Addressed some of the comments from Anirban within this PR.
1. Limit iterations
2. DCHECKs for term advancement
3. voter_distribution as parameter.

The rest of the changes will be made in the next iteration.
